### PR TITLE
Bump PocoCpp to 1.8.1

### PR DESCRIPTION
**Pull request by shekharhimanshu**
Updated hunter-packages: https://github.com/hunter-packages/poco/releases/tag/v1.8.1-p0
Added PocoCpp 1.8.1 and bumped default.

### CODE REVIEW

**Comment by shekharhimanshu** [comment]
>hunter-packages/poco@01b77f9

Fixed them. I will update this PR once hunter-packages/poco@01b77f9 is confirmed.

**Comment by shekharhimanshu** [comment]
Could you review hunter-packages/poco@01b77f9 again?
Here are the fixes, https://github.com/hunter-packages/poco/commit/bf45733463b8f49214e9c81fa6e284a052fa7a14 and https://github.com/hunter-packages/poco/commit/65378773358c2898027857c4c6aefb55850050b0

**Comment by Ubiquite** [comment]
I can help if there is more fixes to do. I'm waiting this PR too

**Comment by shekharhimanshu** [comment]
Hi @ruslo 

Sorry for the delay.
I have fixed issues pointed by you here, https://github.com/hunter-packages/poco/commit/65378773358c2898027857c4c6aefb55850050b0

Could you check them? :)

**Comment by shekharhimanshu** [comment]
@Ubiquite 

>I can help if there is more fixes to do. I'm waiting this PR too

Thanks. 
Please follow hunter-packages/poco review here, https://github.com/hunter-packages/poco/commit/65378773358c2898027857c4c6aefb55850050b0

**Comment by ruslo** [comment]
> Could you review hunter-packages/poco@01b77f9 again?

@shekharhimanshu If you open pull request it will be easier to do review and unreviewed code will never reach `hunter-X.Y.Z` working branch.

**Comment by ciband** [comment]
Waiting on this as well.  Thanks for doing this.

**Comment by shekharhimanshu** [comment]
>If you open pull request it will be easier to do review and unreviewed code will never reach hunter-X.Y.Z working branch.

Yes, I agree and I dont like the current method as well for reviews.
I had no idea that we would be making so many changes to the original Poco project to make it fit to be used with Hunter.
I thought "hunterizing" Poco would be as easy as cherry-pick ing few commits [as you said before](https://github.com/ruslo/hunter/pull/798#issuecomment-302214250). Maybe we need a way to know if some library is "hunter-fit" or not inside hunter-packages.

Anyways, I will open a new PR inside hunter-packages/poco.

**Comment by shekharhimanshu** [comment]
>Anyways, I will open a new PR inside hunter-packages/poco.

https://github.com/hunter-packages/poco/pull/1

**Comment by ruslo** [comment]
> I thought "hunterizing" Poco would be as easy as cherry-pick ing few commits as you said before

You need more libraries hence need to add more `hunter_add_package` calls and `foo::foo` imported libraries integrations.

> Maybe we need a way to know if some library is "hunter-fit" or not inside hunter-packages.

What do you mean?

**Comment by shekharhimanshu** [comment]
>What do you mean?

When I was preparing to bump Poco to 1.7.9, [here's how I hunterized sqlite and zlib](https://github.com/hunter-packages/poco/commit/1d54098e8a9f4f02c5f12f8ec69ffa176d34f6cb) dependencies inside PocoCpp and [it passed all tests and was released](https://github.com/ruslo/hunter/pull/1083). Note that the problem we are trying to correct in https://github.com/hunter-packages/poco/pull/1 was present in PocoCpp v1.7.9-p0 as well but it passed all tests. I did face build error in one scenario and tried to correct it with [v1.7.9-p1](https://github.com/ruslo/hunter/pull/1192).

This time luckily you reviewed the changes and here we are now correcting those. But I am not sure if these are the only changes required and hence wanted some kind of tests in CI to assert that the said library can be used with Hunter. (since the current tests in CI have failed to report such issues in the past)

Another way, of course it to solve problems as they come (what we are doing right now).

**Comment by ruslo** [comment]
> here's how I hunterized sqlite and zlib dependencies inside PocoCpp

Without REQUIRED if sqlite3 not found the missing package will be simply ignored. Without CONFIG in ZLIB there is a chance that you will find system library instead of one from Hunter. You should always check documentation for `hunter_add_package`/`find_package`/imported targets consistency. And even if it will work the final result will not be relocatable because `SQLITE3_LIBRARIES` and `ZLIB_LIBRARIES` is a hardcoded full paths as far as I understand.

> it passed all tests and was released

There is no deep analysis while testing (at least for now) so it means that binaries and `*Config.cmake` with issues can pass the test. As for release I don't check the sources of packages too, it's a package maintainer responsibility. If you not doing some completely insane stuff I will accept pull request.

> This time luckily you reviewed the changes and here we are now correcting those

Yep.

> hence wanted some kind of tests in CI to assert that the said library can be used with Hunter

I think it's possible to do but require more advanced testing system than Hunter currently have. To detect bundled ZLIB version you have to build and install ZLIB package and save the list of symbols somewhere (nm --defined-only --demangle) then compare this list with list of symbols in PocoCpp and report interception.

**Comment by shekharhimanshu** [comment]
Updated to use `hunter-packages/poco v1.8.1-p1` release instead.

I was thinking about squashing https://github.com/ruslo/hunter/pull/1322/commits/414de0f35b9c26522b17d86f911f467c7cbda4ad and  https://github.com/ruslo/hunter/pull/1322/commits/9371a7a781f07d1825b5b8e3f0c35fcc0e043b43 while retaining https://github.com/ruslo/hunter/pull/1322/commits/24181de4f863e5e0b78096d262aa27ff9ced9108.

What do you think?

**Comment by shekharhimanshu** [comment]
I agree with most of your points there and do understand that it will be nearly impossible for you to maintain each and every library that Hunter supports. So it has to be done collectively.

>There is no deep analysis while testing (at least for now) so it means that binaries and *Config.cmake with issues can pass the test. 

How about we try to build the respective example project inside Travis as well? Or are we doing it already?

Cheers!

**Comment by shekharhimanshu** [comment]
>Sounds good.

Done.

>What kind of example? The thing with such type of errors is that they will not be detected while compiling and linking - you have to improve the testing system.

I was talking about [this](https://github.com/ruslo/hunter/tree/master/examples/PocoCpp) example project.

File: cmake/configs/default.cmake [hunk 1]
```diff
@@ -105,7 +105,7 @@ hunter_config(OpenCV-Extra VERSION 3.4.0)
 hunter_config(OpenNMTTokenizer VERSION 0.2.0-p1)
 hunter_config(OpenSSL VERSION 1.1.0g)
 hunter_config(PNG VERSION 1.6.26-p1)
-hunter_config(PocoCpp VERSION 1.7.9-p1)
+hunter_config(PocoCpp VERSION 1.8.1-p0)
```
Comment (ruslo) [inline cmake/configs/default.cmake]: I think it should be `1.8.1-p1`

File: cmake/projects/PocoCpp/hunter.cmake [hunk 2]
```diff
@@ -4,6 +4,17 @@ include(hunter_pick_scheme)
 include(hunter_cacheable)
 include(hunter_download)
 
+hunter_add_version(
+    PACKAGE_NAME
+    PocoCpp
+    VERSION
+    1.8.1-p0
```
Comment (ruslo) [inline cmake/projects/PocoCpp/hunter.cmake]: Real tag mismatch, should be `1.8.1-p1`.

**Comment by ruslo** [comment]
> I was talking about this example project

Not sure I follow, you want to add new code to this example?

File: cmake/projects/PocoCpp/hunter.cmake (hunk 2 above)
Comment (shekharhimanshu) [reply cmake/projects/PocoCpp/hunter.cmake]: Oh, we never released `1.8.1-p0`, so I thought the first release has `p0` and then so on.
Is there a naming rule or something?

Comment (ruslo) [reply cmake/projects/PocoCpp/hunter.cmake]: > Is there a naming rule or something?

I think it's clean and convenient approach when upstream version matches version in Hunter. What is the point of introducing different version?

Comment (shekharhimanshu) [reply cmake/projects/PocoCpp/hunter.cmake]: I understand your point but if we release `1.8.1-p1`, someone might search for `1.8.1-p0` (which we will be skipping). Hunter-packages poco versions are after all internal and abstracted from users using Hunter.

I am good with upstream version <-> Hunter version as well. Will do and squash.

**Comment by shekharhimanshu** [comment]
>Not sure I follow, you want to add new code to this example?

No. I just meant that if we try to build the library example project, we can be at least sure that `find_package(Lib)` will work and that users using it will be able to build and link.

File: cmake/projects/PocoCpp/hunter.cmake (hunk 2 above)
Comment (shekharhimanshu) [reply cmake/projects/PocoCpp/hunter.cmake]: >I am good with upstream version <-> Hunter version as well. Will do and squash.

Done

**Comment by ruslo** [comment]
> if we try to build the library example project, we can be at least sure that find_package(Lib) will work and that users using it will be able to build and link

That's what currently is testing already: `find_package(Poco REQUIRED Foundation Util CONFIG)` and `target_link_libraries(hello-poco Poco::Foundation Poco::Util)` - compile and link.

**Comment by ruslo** [comment]
Testing:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.3369
* https://travis-ci.org/ingenue/hunter/builds/346797465?utm_source=github_status&utm_medium=notification

**Comment by ruslo** [comment]
> Testing

Some tests failed. Please check this page for remote testing instructions:
* https://docs.hunter.sh/en/latest/creating-new/update.html

**Comment by shekharhimanshu** [comment]
>Some tests failed. Please check this page for remote testing instructions:

I think I understand what the problem is. Since we made `POCO_UNBUNDLED=ON`, Poco is trying to find external `PCRE` and failing on toolchains which doesnt have pcre installed on them by default as system libs (android or windows; toolchains like ubuntu are passing the test because they are able to find PCRE installed in the system)

https://github.com/hunter-packages/poco/blob/hunter/Foundation/CMakeLists.txt#L38

I think maybe we should hunterize [PCRE](https://docs.hunter.sh/en/latest/packages/pkg/libpcre.html) as well just like we did for SQLite3 and ZLIB?

**Comment by ruslo** [comment]
> I think maybe we should hunterize PCRE as well just like we did for SQLite3 and ZLIB?

Yes

**Comment by shekharhimanshu** [comment]
>Yes

Alright give me sometime. I would like to test it myself before submitting any new PRs.

**Comment by shekharhimanshu** [comment]
>shekharhimanshu dismissed ruslo’s stale review via 45fb8ca 2 minutes ago

Had to squash commit to keep it clean.

**Comment by shekharhimanshu** [comment]
@ruslo 

I have patched the errors. Could you please test?

I tried starting remote tests by myself following https://docs.hunter.sh/en/latest/creating-new/update.html#testing but couldnt succeed.

From my hunter fork, I did the following:
```
git fetch ci
git checkout pkg.pococpp
git checkout -b test.pococpp
git merge bump-poco-1.8.1
git push -u origin test.pococpp
```

**Comment by ruslo** [comment]
> From my hunter fork, I did the following

Looks good, what kind of a error are you experiencing?

**Comment by shekharhimanshu** [comment]
>Looks good, what kind of a error are you experiencing?

I couldn't find my test runs on either travis or appveyor.

**Comment by ruslo** [comment]
> I couldn't find my test runs on either travis or appveyor

Probably because you haven't registered them?
* https://docs.hunter.sh/en/latest/creating-new/create/cmake.html#ci-testing
```
Register your Hunter fork:
...
```

**Comment by shekharhimanshu** [comment]
@ruslo

Travis [failed](https://travis-ci.org/shekharhimanshu/hunter/jobs/349145564) to missing EXPAT library. PR: https://github.com/hunter-packages/poco/pull/3
AppVeyor [failed](https://ci.appveyor.com/project/shekharhimanshu/hunter/build/job/sqg6vr6v6npou3kn) while building `libpcre`:

```
 -- [hunter *** DEBUG *** 2018-03-05T05:32:46] Settings (initialize):
520  -- [hunter *** DEBUG *** 2018-03-05T05:32:46]   HunterGate done (YES)
521  -- [hunter *** DEBUG *** 2018-03-05T05:32:46]   Cache init (YES)
522  -- [hunter *** DEBUG *** 2018-03-05T05:32:46] Scheme: url_sha1_autotools
523
524  [hunter ** FATAL ERROR **] Autotools scheme not supported with Visual Studio
525  [hunter ** FATAL ERROR **] [Directory:C:/projects/hunter/_testing/Hunter/_Base/49d4946/a9a641a/de5237a/Build/libpcre]
526
527  ------------------------------ WIKI -------------------------------
528      https://github.com/ruslo/hunter/wiki/error.incorrect.input.data
529  -------------------------------------------------------------------
530
```

**Comment by ruslo** [comment]
> Suggestions?

Fix it. E.g. by adding CMake as a build system.

**Comment by ciband** [comment]
Is there anything I could do to help this PR?  I would very much like to use Poco >= 1.8.0 for my project. to get some of the newly added crypto functions  

Thank you.

**Comment by ruslo** [comment]
> Is there anything I could do to help this PR?

@ciband See:

> > AppVeyor failed while building libpcre:
> > It looks like libpcre picks autotools scheme by default which is not supported with Visual Studio.
> > https://github.com/ruslo/hunter/blob/master/cmake/projects/libpcre/hunter.cmake#L25
> > Suggestions?

> Fix it. E.g. by adding CMake as a build system.

I.e. either fix libpcre on Windows or add an option to Poco to exclude libpcre part from build.

**Comment by Bjoe** [comment]
Hi @all ... sorry I will use your pull request to give just a side information: I currently working on to add hunter support into Poco (based on develop version 1.9.1). See [feature-hunter-support branch in githup.com/Bjoe/poco](https://github.com/Bjoe/poco/tree/feature-hunter-support ). 
I'm not so familiar with the license model and I still wanted to ask (my plan was normaly via email) @ruslo if it is ok, that "HunterGate.cmake" is commited/included in Poco/cmake directory? Poco is published with Boost license. Is this a problem?
For the information, since March 2018 I'm a contributor of cmake build system for poco. That poco include hunter support is **not yet an official statement** from Poco project !!! At the moment, it as a personal idea and wish from my side. I should first start a review/pull request so that I get an OK from the maintainers !

**Comment by ruslo** [comment]
> I'm not so familiar with the license model and I still wanted to ask (my plan was normaly via email) @ruslo if it is ok, that "HunterGate.cmake" is commited/included in Poco/cmake directory? Poco is published with Boost license. Is this a problem?

I think it should not be a problem, as far as I know BSD license is quite permissive (that's why it was chosen), probably you should ask Poco developers.

**Comment by mchiasson** [comment]
would be nice if `ENABLE_NETSSL=OFF` ws changed to `ENABLE_NETSSL=ON`, if that works...

**Comment by ruslo** [comment]
Closing inactive pull request.

**Pull request closed without merging**